System objects now appear on template-specified tracks

asattely opened the pull request:
Templates include information about which staves have system objects above them (i.e. tempo text, voltas, jump markers, system text, etc.), and this PR enforces those objects on the appropriate staves.
![image](https://user-images.githubusercontent.com/89263931/145921945-2c109274-eeee-4865-861b-ad1cd2293da1.png)

There is also pre-existing functionality where you can hold ctrl when dragging a volta to a staff and it shows on that staff instead of the preset system staves:
![image](https://user-images.githubusercontent.com/89263931/145922196-108b3280-e8fe-4e90-bb41-68e5a269f136.png)

This PR also deals with saving and loading of all system objects, regardless of staff.

asattely commented:
Bah. Killed the utests. I'll give it a look tomorrow

bkunda commented:
This is looking super nice so far! Thanks so much @asattely 😀

Not sure if you're already onto this (and so apologies if I'm pointing out something you already know!): System objects don't seem to be respecting changes made to top-staff instruments from the Instruments dialog (after the template has been opened). 

In this example:

- Adding a piccolo above the flute: system objects stay with the flute and don't move up a staff
- Removing the violin 1, then re-adding it: system objects move down to violin 2 in first instance (expected), but don't jump back up again to the newly-added violin 1
- Lastly, my intuition told me that I might be able to adjust the bracket around the string staves, in the hopes I could "trick" the system objects to jumping back up. I was being a tad optimistic, I think 😅

![system-objects-1](https://user-images.githubusercontent.com/86290556/145943248-9892379c-56d2-48fb-a48d-899dbc3754d0.gif)

wizofaus commented:
Ah, glad to see someone was brave enough to take this on, I'd looked into it a bit but got bogged down with other stuff - good work though!

njvdberg commented:
@bkunda Since this feature makes use of the score ordering the points you mention are expected. If you add a piccolo the score order is changed to `Custom`. My guess is every is fine after you change the score order back to `Orchestral`. This is related to #9210 and other related issues I'm looking at right now.

@asattely I was working on this a year ago but got distracted. Good to see you picked this up :+1:.

asattely commented:
So, the unit test is failing on split measures, and apparently the tempo isn't being written into the files being diff'd with the reference score. I am having a lot of trouble reproducing this in any scores I create myself. I'm going to keep at this tomorrow to see if there's anything weird I'm missing. The rest of the utests have been fixed, and the vtests magically started being okay again.

asattely commented:
I've added some xml to the save file so that scores can save their list of system object staves. This will be useful in the short-term if we want to make more fine adjustments to templates, and later on when we implement configuration of this stuff.

Unfortunately since it modifies the mscx file slightly, I expect every utest to fail now, so there's going to be a lot of fixing of reference files in my future.

asattely commented:
The utest I was worried about before seems to have fixed itself with a rebase, though I seem to have introduced a new problem with the MM rests which I will investigate now.

asattely commented:
That'll probably do it! vtests are passing, the utests passed last time so I am confident in those, and we should be good for some good old fashioned testin'

Jojo-Schmitz commented:
The vtests fail at multimeasure rests, interrupted by a system element (here: rehearsal mark or time signature), by loosing those.

asattely commented:
Fixed one more bug with jump markers appearing twice in an excerpt, and rebased. As long as all the checks pass, which they should, this one will be ready to gooooo
